### PR TITLE
kicad-testing: init at 7.0-2024-01-07

### PR DESCRIPTION
**Pull request by evils**
update update.sh to include it in versions.nix

fix the update.sh to allow r-ryantm's automatic updates

sort the package inputs a bit

## Description of changes

upstream has been encouraging the use of a testing package to test the changes on the stable branch
(that is, changes that have been backported to stable and will be in the next stable minor or patch release)

This adds such a package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by evils** [comment]
resolved merge conflict

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 255271` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>kicad-testing</li>
    <li>kicad-testing-small</li>
  </ul>
</details>

Seems to run fine on my end, just a couple of nits

File: pkgs/applications/science/electronics/kicad/base.nix [hunk 1]
```diff
@@ -56,6 +57,8 @@
 
 assert lib.assertMsg (!(sanitizeAddress && sanitizeThreads))
   "'sanitizeAddress' and 'sanitizeThreads' are mutually exclusive, use one.";
+assert testing -> !stable
+  -> throw "testing implies stable and cannot be used with stable = false";
```
Comment (evils) [inline pkgs/applications/science/electronics/kicad/base.nix:61]: @pbsds i test for testing / stable mutual exclusivity here
fairly sure i confirmed this works as intended

File: pkgs/applications/science/electronics/kicad/update.sh [hunk 2]
```diff
@@ -97,32 +115,56 @@ printf "{\n"
...
   # skip a version if we don't want to check it
-  if [[ (${version} != "master" && -n ${check_stable}) \
-     || (${version} == "master" && -n ${check_unstable}) ]]; then
+  if [[ (-n ${check_stable} && ${version} != "master" && ${version} != "testing") \
+     || (-n ${check_testing} && ${version} == "testing") \
+     || (-n ${check_unstable} && ${version} == "master" ) ]]; then
+
+    if [[ ${version} == "master" ]]; then
+      pname="kicad-unstable"
+      new_version="${now}"
```
[the remark this answers is not in the document]
Comment (evils) [reply pkgs/applications/science/electronics/kicad/update.sh:143]: probably, but i think that could result in some wildly varying dates if the latest commit is one authored long ago (if for example an old merge request was just merged)

File: pkgs/applications/science/electronics/kicad/default.nix [hunk 3]
```diff
@@ -75,7 +78,9 @@
 # }
 
 let
-  baseName = if (stable) then "kicad" else "kicad-unstable";
+  baseName = if (testing) then "kicad-testing"
+    else if (stable) then "kicad"
+    else "kicad-unstable";
```
[the remark this answers is not in the document]
Comment (evils) [reply pkgs/applications/science/electronics/kicad/default.nix:83]: i test that in base.nix, pinged you over there

**Comment by evils** [comment]
@pbsds have i resolved your nits?

**Comment by pbsds** [comment]
Thanks for the ping, somehow marked this as read. See https://github.com/NixOS/nixpkgs/pull/255271#discussion_r1406476500

File: pkgs/applications/science/electronics/kicad/update.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (evils) [reply pkgs/applications/science/electronics/kicad/update.sh:149]: implemented, seems to work as intended, thanks!

and sorry for such a late response

**Comment by evils** [comment]
added the use of the `create_at` field from gitlab
updated the testing package
and tested the updated package

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 255271` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>kicad-testing</li>
    <li>kicad-testing-small</li>
  </ul>
</details>

Tested the update script, which seems to be working. `nix-update -u -f nixpkgs/ kicad-testing` however fails with `fatal: not a git repository`, but `nix-update -u kicad-testing`  works when workdir is nixpkgs root. I consider that issue out of scope for this PR.

LGTM

**Pull request merged**